# net-snmp 5.9.5.2 reads ASN1_STRING fields directly, which OpenSSL 4.0 made opaque;
# the public accessors work on every OpenSSL version it supports.
--- a/snmplib/snmp_openssl.c	2026-10-16 18:45:19.941959909 +0700
+++ b/snmplib/snmp_openssl.c	2026-10-16 18:45:20.039013213 +0700
@@ -253,7 +253,7 @@
         netsnmp_assert(NULL != oname_entry);
         oname_value = X509_NAME_ENTRY_get_data(oname_entry);
 
-        if (oname_value->type != V_ASN1_PRINTABLESTRING)
+        if (ASN1_STRING_type(oname_value) != V_ASN1_PRINTABLESTRING)
             continue;
 
         /** get NID */
@@ -268,7 +268,7 @@
 
         DEBUGMSGT(("9:cert:dump:names",
                    "[%02d] NID type %d, ASN type %d\n", i, onid,
-                   oname_value->type));
+                   ASN1_STRING_type(oname_value)));
         DEBUGMSGT(("9:cert:dump:names", "%s/%s: '%s'\n", prefix_long,
                    prefix_short, ASN1_STRING_get0_data(oname_value)));
     }
@@ -434,19 +434,19 @@
             break;
 
         case GEN_IPADD:
-            if (oname->d.iPAddress->length == 4) {
-                sprintf(ipbuf, "%d.%d.%d.%d", oname->d.iPAddress->data[0],
-                        oname->d.iPAddress->data[1],
-                        oname->d.iPAddress->data[2],
-                        oname->d.iPAddress->data[3]);
+            if (ASN1_STRING_length(oname->d.iPAddress) == 4) {
+                sprintf(ipbuf, "%d.%d.%d.%d", ASN1_STRING_get0_data(oname->d.iPAddress)[0],
+                        ASN1_STRING_get0_data(oname->d.iPAddress)[1],
+                        ASN1_STRING_get0_data(oname->d.iPAddress)[2],
+                        ASN1_STRING_get0_data(oname->d.iPAddress)[3]);
                 rtn = strdup(ipbuf);
             }
-            else if ((oname->d.iPAddress->length == 16) ||
-                     (oname->d.iPAddress->length == 20)) {
+            else if ((ASN1_STRING_length(oname->d.iPAddress) == 16) ||
+                     (ASN1_STRING_length(oname->d.iPAddress) == 20)) {
                 char *pos = ipbuf;
                 int   j;
-                for(j = 0; j < oname->d.iPAddress->length; ++j) {
-                    *pos++ = VAL2HEX(oname->d.iPAddress->data[j]);
+                for(j = 0; j < ASN1_STRING_length(oname->d.iPAddress); ++j) {
+                    *pos++ = VAL2HEX(ASN1_STRING_get0_data(oname->d.iPAddress)[j]);
                     *pos++ = ':';
                 }
                 *pos = '\0';
@@ -454,7 +454,7 @@
             }
             else
                 NETSNMP_LOGONCE((LOG_WARNING, "unexpected ip addr length %d\n",
-                       oname->d.iPAddress->length));
+                       ASN1_STRING_length(oname->d.iPAddress)));
 
             break;
         default:
